File: modules/angular2/src/core/change_detection/parser/ast.ts
import {isBlank, isPresent, FunctionWrapper, BaseException} from "angular2/src/core/facade/lang";
import {Map, ListWrapper, StringMapWrapper} from "angular2/src/core/facade/collection";

export class AST {
  visit(visitor: AstVisitor): any { return null; }
  toString(): string { return "AST"; }
}

export class EmptyExpr extends AST {
  visit(visitor: AstVisitor) {
    // do nothing
  }
}

export class ImplicitReceiver extends AST {
  visit(visitor: AstVisitor): any { return visitor.visitImplicitReceiver(this); }
}

/**
 * Multiple expressions separated by a semicolon.
 */
export class Chain extends AST {
  constructor(public expressions: any[]) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitChain(this); }
}

export class Conditional extends AST {
  constructor(public condition: AST, public trueExp: AST, public falseExp: AST) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitConditional(this); }
}

export class If extends AST {
  constructor(public condition: AST, public trueExp: AST, public falseExp?: AST) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitIf(this); }
}

export class PropertyRead extends AST {
  constructor(public receiver: AST, public name: string, public getter: Function) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitPropertyRead(this); }
}

export class PropertyWrite extends AST {
  constructor(public receiver: AST, public name: string, public setter: Function,
              public value: AST) {
    super();
  }
  visit(visitor: AstVisitor): any { return visitor.visitPropertyWrite(this); }
}

export class SafePropertyRead extends AST {
  constructor(public receiver: AST, public name: string, public getter: Function) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitSafePropertyRead(this); }
}

export class KeyedRead extends AST {
  constructor(public obj: AST, public key: AST) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitKeyedRead(this); }
}

export class KeyedWrite extends AST {
  constructor(public obj: AST, public key: AST, public value: AST) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitKeyedWrite(this); }
}

export class BindingPipe extends AST {
  constructor(public exp: AST, public name: string, public args: any[]) { super(); }

  visit(visitor: AstVisitor): any { return visitor.visitPipe(this); }
}

export class LiteralPrimitive extends AST {
  constructor(public value) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitLiteralPrimitive(this); }
}

export class LiteralArray extends AST {
  constructor(public expressions: any[]) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitLiteralArray(this); }
}

export class LiteralMap extends AST {
  constructor(public keys: any[], public values: any[]) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitLiteralMap(this); }
}

export class Interpolation extends AST {
  constructor(public strings: any[], public expressions: any[]) { super(); }
  visit(visitor: AstVisitor) { visitor.visitInterpolation(this); }
}

export class Binary extends AST {
  constructor(public operation: string, public left: AST, public right: AST) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitBinary(this); }
}

export class PrefixNot extends AST {
  constructor(public expression: AST) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitPrefixNot(this); }
}

export class MethodCall extends AST {
  constructor(public receiver: AST, public name: string, public fn: Function, public args: any[]) {
    super();
  }
  visit(visitor: AstVisitor): any { return visitor.visitMethodCall(this); }
}

export class SafeMethodCall extends AST {
  constructor(public receiver: AST, public name: string, public fn: Function, public args: any[]) {
    super();
  }
  visit(visitor: AstVisitor): any { return visitor.visitSafeMethodCall(this); }
}

export class FunctionCall extends AST {
  constructor(public target: AST, public args: any[]) { super(); }
  visit(visitor: AstVisitor): any { return visitor.visitFunctionCall(this); }
}

export class ASTWithSource extends AST {
  constructor(public ast: AST, public source: string, public location: string) { super(); }
  visit(visitor: AstVisitor): any { return this.ast.visit(visitor); }
  toString(): string { return `${this.source} in ${this.location}`; }
}

export class TemplateBinding {
  constructor(public key: string, public keyIsVar: boolean, public name: string,
              public expression: ASTWithSource) {}
}

export interface AstVisitor {
  visitBinary(ast: Binary): any;
  visitChain(ast: Chain): any;
  visitConditional(ast: Conditional): any;
  visitFunctionCall(ast: FunctionCall): any;
  visitIf(ast: If): any;
  visitImplicitReceiver(ast: ImplicitReceiver): any;
  visitInterpolation(ast: Interpolation): any;
  visitKeyedRead(ast: KeyedRead): any;
  visitKeyedWrite(ast: KeyedWrite): any;
  visitLiteralArray(ast: LiteralArray): any;
  visitLiteralMap(ast: LiteralMap): any;
  visitLiteralPrimitive(ast: LiteralPrimitive): any;
  visitMethodCall(ast: MethodCall): any;
  visitPipe(ast: BindingPipe): any;
  visitPrefixNot(ast: PrefixNot): any;
  visitPropertyRead(ast: PropertyRead): any;
  visitPropertyWrite(ast: PropertyWrite): any;
  visitSafeMethodCall(ast: SafeMethodCall): any;
  visitSafePropertyRead(ast: SafePropertyRead): any;
}

export class RecursiveAstVisitor implements AstVisitor {
  visitBinary(ast: Binary): any {
    ast.left.visit(this);
    ast.right.visit(this);
    return null;
  }
  visitChain(ast: Chain): any { return this.visitAll(ast.expressions); }
  visitConditional(ast: Conditional): any {
    ast.condition.visit(this);
    ast.trueExp.visit(this);
    ast.falseExp.visit(this);
    return null;
  }
  visitIf(ast: If): any {
    ast.condition.visit(this);
    ast.trueExp.visit(this);
    ast.falseExp.visit(this);
    return null;
  }
  visitPipe(ast: BindingPipe): any {
    ast.exp.visit(this);
    this.visitAll(ast.args);
    return null;
  }
  visitFunctionCall(ast: FunctionCall): any {
    ast.target.visit(this);
    this.visitAll(ast.args);
    return null;
  }
  visitImplicitReceiver(ast: ImplicitReceiver): any { return null; }
  visitInterpolation(ast: Interpolation): any { return this.visitAll(ast.expressions); }
  visitKeyedRead(ast: KeyedRead): any {
    ast.obj.visit(this);
    ast.key.visit(this);
    return null;
  }
  visitKeyedWrite(ast: KeyedWrite): any {
    ast.obj.visit(this);
    ast.key.visit(this);
    ast.value.visit(this);
    return null;
  }
  visitLiteralArray(ast: LiteralArray): any { return this.visitAll(ast.expressions); }
  visitLiteralMap(ast: LiteralMap): any { return this.visitAll(ast.values); }
  visitLiteralPrimitive(ast: LiteralPrimitive): any { return null; }
  visitMethodCall(ast: MethodCall): any {
    ast.receiver.visit(this);
    return this.visitAll(ast.args);
  }
  visitPrefixNot(ast: PrefixNot): any {
    ast.expression.visit(this);
    return null;
  }
  visitPropertyRead(ast: PropertyRead): any {
    ast.receiver.visit(this);
    return null;
  }
  visitPropertyWrite(ast: PropertyWrite): any {
    ast.receiver.visit(this);
    ast.value.visit(this);
    return null;
  }
  visitSafePropertyRead(ast: SafePropertyRead): any {
    ast.receiver.visit(this);
    return null;
  }
  visitSafeMethodCall(ast: SafeMethodCall): any {
    ast.receiver.visit(this);
    return this.visitAll(ast.args);
  }
  visitAll(asts: AST[]): any {
    ListWrapper.forEach(asts, (ast) => { ast.visit(this); });
    return null;
  }
}

export class AstTransformer implements AstVisitor {
  visitImplicitReceiver(ast: ImplicitReceiver): ImplicitReceiver { return ast; }

  visitInterpolation(ast: Interpolation): Interpolation {
    return new Interpolation(ast.strings, this.visitAll(ast.expressions));
  }

  visitLiteralPrimitive(ast: LiteralPrimitive): LiteralPrimitive {
    return new LiteralPrimitive(ast.value);
  }

  visitPropertyRead(ast: PropertyRead): PropertyRead {
    return new PropertyRead(ast.receiver.visit(this), ast.name, ast.getter);
  }

  visitPropertyWrite(ast: PropertyWrite): PropertyWrite {
    return new PropertyWrite(ast.receiver.visit(this), ast.name, ast.setter, ast.value);
  }

  visitSafePropertyRead(ast: SafePropertyRead): SafePropertyRead {
    return new SafePropertyRead(ast.receiver.visit(this), ast.name, ast.getter);
  }

  visitMethodCall(ast: MethodCall): MethodCall {
    return new MethodCall(ast.receiver.visit(this), ast.name, ast.fn, this.visitAll(ast.args));
  }

  visitSafeMethodCall(ast: SafeMethodCall): SafeMethodCall {
    return new SafeMethodCall(ast.receiver.visit(this), ast.name, ast.fn, this.visitAll(ast.args));
  }

  visitFunctionCall(ast: FunctionCall): FunctionCall {
    return new FunctionCall(ast.target.visit(this), this.visitAll(ast.args));
  }

  visitLiteralArray(ast: LiteralArray): LiteralArray {
    return new LiteralArray(this.visitAll(ast.expressions));
  }

  visitLiteralMap(ast: LiteralMap): LiteralMap {
    return new LiteralMap(ast.keys, this.visitAll(ast.values));
  }

  visitBinary(ast: Binary): Binary {
    return new Binary(ast.operation, ast.left.visit(this), ast.right.visit(this));
  }

  visitPrefixNot(ast: PrefixNot): PrefixNot { return new PrefixNot(ast.expression.visit(this)); }

  visitConditional(ast: Conditional): Conditional {
    return new Conditional(ast.condition.visit(this), ast.trueExp.visit(this),
                           ast.falseExp.visit(this));
  }

  visitPipe(ast: BindingPipe): BindingPipe {
    return new BindingPipe(ast.exp.visit(this), ast.name, this.visitAll(ast.args));
  }

  visitKeyedRead(ast: KeyedRead): KeyedRead {
    return new KeyedRead(ast.obj.visit(this), ast.key.visit(this));
  }

  visitKeyedWrite(ast: KeyedWrite): KeyedWrite {
    return new KeyedWrite(ast.obj.visit(this), ast.key.visit(this), ast.value.visit(this));
  }

  visitAll(asts: any[]): any[] {
    var res = ListWrapper.createFixedSize(asts.length);
    for (var i = 0; i < asts.length; ++i) {
      res[i] = asts[i].visit(this);
    }
    return res;
  }

  visitChain(ast: Chain): Chain { return new Chain(this.visitAll(ast.expressions)); }

  visitIf(ast: If): If {
    let falseExp = isPresent(ast.falseExp) ? ast.falseExp.visit(this) : null;
    return new If(ast.condition.visit(this), ast.trueExp.visit(this), falseExp);
  }
}

File: lib/src/core/change_detection/parser/ast.dart
library angular2.src.core.change_detection.parser.ast;

import "package:angular2/src/facade/collection.dart" show ListWrapper;

class AST {
  dynamic visit(AstVisitor visitor) {
    return null;
  }

  String toString() {
    return "AST";
  }
}

class EmptyExpr extends AST {
  visit(AstVisitor visitor) {}
}

class ImplicitReceiver extends AST {
  dynamic visit(AstVisitor visitor) {
    return visitor.visitImplicitReceiver(this);
  }
}

/**
 * Multiple expressions separated by a semicolon.
 */
class Chain extends AST {
  List<dynamic> expressions;
  Chain(this.expressions) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitChain(this);
  }
}

class Conditional extends AST {
  AST condition;
  AST trueExp;
  AST falseExp;
  Conditional(this.condition, this.trueExp, this.falseExp) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitConditional(this);
  }
}

class PropertyRead extends AST {
  AST receiver;
  String name;
  Function getter;
  PropertyRead(this.receiver, this.name, this.getter) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitPropertyRead(this);
  }
}

class PropertyWrite extends AST {
  AST receiver;
  String name;
  Function setter;
  AST value;
  PropertyWrite(this.receiver, this.name, this.setter, this.value) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitPropertyWrite(this);
  }
}

class SafePropertyRead extends AST {
  AST receiver;
  String name;
  Function getter;
  SafePropertyRead(this.receiver, this.name, this.getter) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitSafePropertyRead(this);
  }
}

class KeyedRead extends AST {
  AST obj;
  AST key;
  KeyedRead(this.obj, this.key) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitKeyedRead(this);
  }
}

class KeyedWrite extends AST {
  AST obj;
  AST key;
  AST value;
  KeyedWrite(this.obj, this.key, this.value) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitKeyedWrite(this);
  }
}

class BindingPipe extends AST {
  AST exp;
  String name;
  List<dynamic> args;
  BindingPipe(this.exp, this.name, this.args) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitPipe(this);
  }
}

class LiteralPrimitive extends AST {
  var value;
  LiteralPrimitive(this.value) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitLiteralPrimitive(this);
  }
}

class LiteralArray extends AST {
  List<dynamic> expressions;
  LiteralArray(this.expressions) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitLiteralArray(this);
  }
}

class LiteralMap extends AST {
  List<dynamic> keys;
  List<dynamic> values;
  LiteralMap(this.keys, this.values) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitLiteralMap(this);
  }
}

class Interpolation extends AST {
  List<dynamic> strings;
  List<dynamic> expressions;
  Interpolation(this.strings, this.expressions) : super() {
    /* super call moved to initializer */;
  }
  visit(AstVisitor visitor) {
    visitor.visitInterpolation(this);
  }
}

class Binary extends AST {
  String operation;
  AST left;
  AST right;
  Binary(this.operation, this.left, this.right) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitBinary(this);
  }
}

class PrefixNot extends AST {
  AST expression;
  PrefixNot(this.expression) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitPrefixNot(this);
  }
}

class MethodCall extends AST {
  AST receiver;
  String name;
  Function fn;
  List<dynamic> args;
  MethodCall(this.receiver, this.name, this.fn, this.args) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitMethodCall(this);
  }
}

class SafeMethodCall extends AST {
  AST receiver;
  String name;
  Function fn;
  List<dynamic> args;
  SafeMethodCall(this.receiver, this.name, this.fn, this.args) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitSafeMethodCall(this);
  }
}

class FunctionCall extends AST {
  AST target;
  List<dynamic> args;
  FunctionCall(this.target, this.args) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return visitor.visitFunctionCall(this);
  }
}

class ASTWithSource extends AST {
  AST ast;
  String source;
  String location;
  ASTWithSource(this.ast, this.source, this.location) : super() {
    /* super call moved to initializer */;
  }
  dynamic visit(AstVisitor visitor) {
    return this.ast.visit(visitor);
  }

  String toString() {
    return '''${ this . source} in ${ this . location}''';
  }
}

class TemplateBinding {
  String key;
  bool keyIsVar;
  String name;
  ASTWithSource expression;
  TemplateBinding(this.key, this.keyIsVar, this.name, this.expression) {}
}

abstract class AstVisitor {
  dynamic visitBinary(Binary ast);
  dynamic visitChain(Chain ast);
  dynamic visitConditional(Conditional ast);
  dynamic visitFunctionCall(FunctionCall ast);
  dynamic visitImplicitReceiver(ImplicitReceiver ast);
  dynamic visitInterpolation(Interpolation ast);
  dynamic visitKeyedRead(KeyedRead ast);
  dynamic visitKeyedWrite(KeyedWrite ast);
  dynamic visitLiteralArray(LiteralArray ast);
  dynamic visitLiteralMap(LiteralMap ast);
  dynamic visitLiteralPrimitive(LiteralPrimitive ast);
  dynamic visitMethodCall(MethodCall ast);
  dynamic visitPipe(BindingPipe ast);
  dynamic visitPrefixNot(PrefixNot ast);
  dynamic visitPropertyRead(PropertyRead ast);
  dynamic visitPropertyWrite(PropertyWrite ast);
  dynamic visitSafeMethodCall(SafeMethodCall ast);
  dynamic visitSafePropertyRead(SafePropertyRead ast);
}

class RecursiveAstVisitor implements AstVisitor {
  dynamic visitBinary(Binary ast) {
    ast.left.visit(this);
    ast.right.visit(this);
    return null;
  }

  dynamic visitChain(Chain ast) {
    return this.visitAll(ast.expressions);
  }

  dynamic visitConditional(Conditional ast) {
    ast.condition.visit(this);
    ast.trueExp.visit(this);
    ast.falseExp.visit(this);
    return null;
  }

  dynamic visitPipe(BindingPipe ast) {
    ast.exp.visit(this);
    this.visitAll(ast.args);
    return null;
  }

  dynamic visitFunctionCall(FunctionCall ast) {
    ast.target.visit(this);
    this.visitAll(ast.args);
    return null;
  }

  dynamic visitImplicitReceiver(ImplicitReceiver ast) {
    return null;
  }

  dynamic visitInterpolation(Interpolation ast) {
    return this.visitAll(ast.expressions);
  }

  dynamic visitKeyedRead(KeyedRead ast) {
    ast.obj.visit(this);
    ast.key.visit(this);
    return null;
  }

  dynamic visitKeyedWrite(KeyedWrite ast) {
    ast.obj.visit(this);
    ast.key.visit(this);
    ast.value.visit(this);
    return null;
  }

  dynamic visitLiteralArray(LiteralArray ast) {
    return this.visitAll(ast.expressions);
  }

  dynamic visitLiteralMap(LiteralMap ast) {
    return this.visitAll(ast.values);
  }

  dynamic visitLiteralPrimitive(LiteralPrimitive ast) {
    return null;
  }

  dynamic visitMethodCall(MethodCall ast) {
    ast.receiver.visit(this);
    return this.visitAll(ast.args);
  }

  dynamic visitPrefixNot(PrefixNot ast) {
    ast.expression.visit(this);
    return null;
  }

  dynamic visitPropertyRead(PropertyRead ast) {
    ast.receiver.visit(this);
    return null;
  }

  dynamic visitPropertyWrite(PropertyWrite ast) {
    ast.receiver.visit(this);
    ast.value.visit(this);
    return null;
  }

  dynamic visitSafePropertyRead(SafePropertyRead ast) {
    ast.receiver.visit(this);
    return null;
  }

  dynamic visitSafeMethodCall(SafeMethodCall ast) {
    ast.receiver.visit(this);
    return this.visitAll(ast.args);
  }

  dynamic visitAll(List<AST> asts) {
    asts.forEach((ast) => ast.visit(this));
    return null;
  }
}

class AstTransformer implements AstVisitor {
  ImplicitReceiver visitImplicitReceiver(ImplicitReceiver ast) {
    return ast;
  }

  Interpolation visitInterpolation(Interpolation ast) {
    return new Interpolation(ast.strings, this.visitAll(ast.expressions));
  }

  LiteralPrimitive visitLiteralPrimitive(LiteralPrimitive ast) {
    return new LiteralPrimitive(ast.value);
  }

  PropertyRead visitPropertyRead(PropertyRead ast) {
    return new PropertyRead(ast.receiver.visit(this), ast.name, ast.getter);
  }

  PropertyWrite visitPropertyWrite(PropertyWrite ast) {
    return new PropertyWrite(
        ast.receiver.visit(this), ast.name, ast.setter, ast.value);
  }

  SafePropertyRead visitSafePropertyRead(SafePropertyRead ast) {
    return new SafePropertyRead(ast.receiver.visit(this), ast.name, ast.getter);
  }

  MethodCall visitMethodCall(MethodCall ast) {
    return new MethodCall(
        ast.receiver.visit(this), ast.name, ast.fn, this.visitAll(ast.args));
  }

  SafeMethodCall visitSafeMethodCall(SafeMethodCall ast) {
    return new SafeMethodCall(
        ast.receiver.visit(this), ast.name, ast.fn, this.visitAll(ast.args));
  }

  FunctionCall visitFunctionCall(FunctionCall ast) {
    return new FunctionCall(ast.target.visit(this), this.visitAll(ast.args));
  }

  LiteralArray visitLiteralArray(LiteralArray ast) {
    return new LiteralArray(this.visitAll(ast.expressions));
  }

  LiteralMap visitLiteralMap(LiteralMap ast) {
    return new LiteralMap(ast.keys, this.visitAll(ast.values));
  }

  Binary visitBinary(Binary ast) {
    return new Binary(
        ast.operation, ast.left.visit(this), ast.right.visit(this));
  }

  PrefixNot visitPrefixNot(PrefixNot ast) {
    return new PrefixNot(ast.expression.visit(this));
  }

  Conditional visitConditional(Conditional ast) {
    return new Conditional(ast.condition.visit(this), ast.trueExp.visit(this),
        ast.falseExp.visit(this));
  }

  BindingPipe visitPipe(BindingPipe ast) {
    return new BindingPipe(
        ast.exp.visit(this), ast.name, this.visitAll(ast.args));
  }

  KeyedRead visitKeyedRead(KeyedRead ast) {
    return new KeyedRead(ast.obj.visit(this), ast.key.visit(this));
  }

  KeyedWrite visitKeyedWrite(KeyedWrite ast) {
    return new KeyedWrite(
        ast.obj.visit(this), ast.key.visit(this), ast.value.visit(this));
  }

  List<dynamic> visitAll(List<dynamic> asts) {
    var res = ListWrapper.createFixedSize(asts.length);
    for (var i = 0; i < asts.length; ++i) {
      res[i] = asts[i].visit(this);
    }
    return res;
  }

  Chain visitChain(Chain ast) {
    return new Chain(this.visitAll(ast.expressions));
  }
}
